miner, txpool: detect supervisor failsafe and reject interop transactions if enabled

geoknee opened the pull request:
The supervisor is already consulted (via a CheckAccessList call) on tx ingress into the mempool.

This PR does the following:
* adds a local boolean to the backend to cache whether the supervisor response indicates failsafe is active or inactive
* adds a routine to the miner which refreshes that boolean (by making further RPC calls) on a ticker
* has txs evicted from the mempool if failsafe is detected to be active
* has txs rejected at block building time if failsafe was active last time we checked

Also adds a test for handling interop transactions under different supervisor responses.

geoknee commented on miner/op_interop_miner_test.go:
```diff
@@ -0,0 +1,228 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package miner implements Ethereum block creation and mining.
```
I'm not sure if we need this preamble, b/c this is a net new file that doesn't exist (and won't) upstream...

tynes commented:
This hooks into the `checkInterop` functionality that was being removed as part of https://github.com/ethereum-optimism/op-geth/pull/585

geoknee commented:
> This hooks into the `checkInterop` functionality that was being removed as part of #585

Thanks for highlighting. 

New plan:

* Remove RPC calls from miner altogether
* Have op-geth store an atomic Bool indicating if failsafe mode is active on the supervisor
  - update this on a ticker (separate thread) by making an rpc call to the supervisor
  - update this each time we do an ingress filter check on a tx in the mempool (we are already calling out to the supervisor so this doesn't cost us anything)
  * the miner can inspect the local bool when it checks an interop transaction (no RPC call required)..

geoknee committed:
miner, txpool: detect supervisor failsafe and reject interop transactions if enabled

Add routine to periodically check for failsafe mode by calling supervisor RPC and checking error for sentinel value
Backend caches result, miner inspects the cache. Txpool checks over RPC on ingress.

Adds comprehensive tests.

geoknee committed:
move SupervisorInFailSafe method to BackendWithInterop interface

axelKingsley commented on core/types/interoptypes/interop.go:
```diff
@@ -101,3 +101,5 @@ func TxToInteropAccessList(tx *types.Transaction) []common.Hash {
 	return hashes
 }
+
+var ErrFailsafeEnabled = errors.New("failsafe is enabled, rejecting all CheckAccessList requests")
```
Because geth is not the entity doing the rejecting, maybe something more like "all CheckAcessList requests would be rejected" or something?

geoknee replied:
This is a sentinel error returned by the supervisor which we are pattern matching against, so we can't change it here.

axelKingsley commented on eth/backend.go:
```diff
@@ -690,3 +691,11 @@ func (s *Ethereum) HandleRequiredProtocolVersion(required params.ProtocolVersion
+	s.supervisorFailsafe.Store(enabled)
+}
+
+func (s *Ethereum) SupervisorInFailsafe() bool {
```
to fit the format, maybe GetSupervisorFailsafe?

geoknee replied:
[32167ab](https://github.com/ethereum-optimism/op-geth/pull/636/commits/32167abc3a5ec312d419c4034800bddcbbdbbbeb)

axelKingsley commented on miner/miner.go:
```diff
@@ -119,6 +121,42 @@ func New(eth Backend, config Config, engine consensus.Engine) *Miner {
+			case <-ticker.C:
+				ctx, cancel := context.WithTimeout(miner.lifeCtx, 1*time.Second)
+				defer cancel()
+				// Simply calling CheckAccessList will update the failsafe status in the backend
+				// We do not need to do anything with the result
+				_ = backend.CheckAccessList(ctx, []common.Hash{}, interoptypes.CrossUnsafe, interoptypes.ExecutingDescriptor{})
```
A little odd to use stubbed values here. There is an Admin API `admin_getFailsafeEnabled` that would return the value without having to provide stubbed values.

geoknee replied:
Agreed. Made that change here [68bb732](https://github.com/ethereum-optimism/op-geth/pull/636/commits/68bb73204d0666c22b67408193a292c0c1b436b0). 

The only thing giving me pause is that we now need to use the _admin_ API which seems to suggest elevated privileges. But the getter itself ought not to need elevated privileges.

axelKingsley replied:
I *might* argue that even seeing this internal detail is an administrative RPC, even if it is not doing administrative-level modifications.

geoknee replied:
I mean, the implementation at the moment is such that it _is_ in the admin namespace. And it might be a bit ugly to move the getter out of that namespace. But logically, it doesn't seem to me like it should require privileges to call this getter. It is already a part of the public API because we return a specific error from non-admin API calls which indicates whether failsafe is enabled. 

I think we have three ways forward:
1. infer failsafe indirectly through `CheckAccessList`, stubbing the parameters out
2. infer through the dedicated getter thru the admin API
3. move or replicate the getter in the non-admin API namespace of the supervisor, and query that from op-geth

axelKingsley commented on miner/worker.go:
```diff
@@ -369,6 +372,16 @@ func (miner *Miner) makeEnv(parent *types.Header, header *types.Header, coinbase
 
 func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) error {
+	// OP-Stack addition
+	if len(tx.AccessList()) > 0 {
```
Simply having an AccessList does not imply interop, should use `TxToInteropAccessList` instead

geoknee replied:
Good catch, fixed it here [d62d961](https://github.com/ethereum-optimism/op-geth/pull/636/commits/d62d9617e2998e75331108f385685181bcc4b1e2)

axelKingsley commented on miner/worker.go:
```diff
@@ -55,6 +55,9 @@ var (
 	errBlockInterruptedByTimeout  = errors.New("timeout while building block")
 	errBlockInterruptedByResolve  = errors.New("payload resolution while building block")
 
+	// OP-Stack addition
+	errSupervisorInFailsafe = errors.New("supervisor in failsafe")
+
```
can we use the other error you already defined in the interop package?

geoknee replied:
I don't think so. They are two different types of error: this one is local to op-geth and returned by `miner.commitTransaction`. The other one is defined in op-supervisor and duplicated inside op-geth so that it can pattern match against it (i.e. detect that particular error).

axelKingsley commented on miner/op_interop_miner_test.go:
```diff
@@ -0,0 +1,190 @@
+	return miner, testBankKey, testBankAddress
+}
+
+func createInteropTransaction(t *testing.T, miner *Miner, testBankKey *ecdsa.PrivateKey, testUserAddress common.Address) *types.Transaction {
```
Thanks for going the extra mile and adding these tests!

geoknee committed:
rename

geoknee committed:
use interoptypes.TxToInteropAccessList in commitTransaction

geoknee committed:
use admin_getFailsafeEnabled instead of stubbed CheckAccessList

geoknee committed:
fixes

geoknee committed:
fixes

geoknee commented:
Flagging that we should consider existing periodic checks of the ingress filter, and whether we want to harmonise that with what is added on this PR. Also https://github.com/ethereum-optimism/op-geth/issues/635.

geoknee committed:
Merge remote-tracking branch 'origin/optimism' into gk/check-access-list
